util/promlint/promlint.go

```diff
@@ -205,6 +210,54 @@ func lintHistogramSummaryReserved(mf dto.MetricFamily) []Problem {
 	return problems
 }
 
+// lintMetricTypeInName detects when the type of the metric is included in the metric name.
+func lintMetricTypeInName(mf dto.MetricFamily) []Problem {
+	t := mf.GetType()
+	if t == dto.MetricType_UNTYPED {
+		return nil
+	}
+
+	var problems problems
+	n := strings.ToLower(mf.GetName())
+
+	typename := strings.ToLower(t.String())
+	if strings.Contains(n, "_"+typename+"_") || strings.HasSuffix(n, "_"+typename) {
+		problems.Add(mf, fmt.Sprintf(`%s metrics should not include the type in metric name`, typename))
+	}
+	return problems
+}
+
+// lintReservedChars detects colons in metric names.
+func lintReservedChars(mf dto.MetricFamily) []Problem {
+	var problems problems
+	if strings.Contains(mf.GetName(), ":") {
+		problems.Add(mf, "metric names should not contain ':'")
```

Comment:
metric names can contain :, but not from exporters/direct instrumentation. So it's fine from /federate for example (and indeed should be the case).

Reply:
I completely agree. My thoughts are currently that this tool should be targeted for exporters/direct instrumentation so ':' should fail, but I do not hold that opinion tightly. Perhaps the description should contain more information?

Reply:
@brian-brazil I think this is the only unresolved conversation left. Do you feel strongly that linting should work with `/federate` or is it ok to target exporters/direct instrumentation?

Reply:
It should work with both, but mainly exporters. What we don't want is is continuous bugs filed by users asking why /federate doesn't pass.

Reply:
Right now it looks like /federate will not pass the linting in master due to /federate not presenting help text. Since there are not continuous bugs filed by users right now I don't think it will be a problem, but I can remove the `:` check if you feel better about that.

Reply:
We can leave it in then